demos/common/utilities/jws_using_mbedtls.c

```diff
@@ -0,0 +1,634 @@
+
+    while( ulIndex < ulSignatureLength )
+    {
+        if( *pucSignature == '-' )
+        {
+            *pucSignature = '+';
+        }
+        else if( *pucSignature == '_' )
+        {
+            *pucSignature = '/';
+        }
+
+        pucSignature++;
+        ulIndex++;
+    }
+}
+
+/**
+ * @brief Calculate the SHA256 over a buffer of bytes
+ *
+ * @param pucInput The input buffer over which to calculate the SHA256.
+ * @param ulInputLength The length of \p pucInput.
+ * @param pucOutput The output buffer into which the SHA256. It must be 32 bytes in length.
+ * @return uint32_t The result of the operation.
+ * @retval 0 if successful.
+ * @retval Non-0 if not successful.
+ */
+static uint32_t prvJWS_SHA256Calculate( const unsigned char * pucInput,
+                                        uint32_t ulInputLength,
+                                        unsigned char * pucOutput )
+{
+    mbedtls_md_context_t ctx;
+    mbedtls_md_type_t md_type = MBEDTLS_MD_SHA256;
+
+    mbedtls_md_init( &ctx );
+    mbedtls_md_setup( &ctx, mbedtls_md_info_from_type( md_type ), 0 );
+    mbedtls_md_starts( &ctx );
+    mbedtls_md_update( &ctx, pucInput, ulInputLength );
+    mbedtls_md_finish( &ctx, pucOutput );
+    mbedtls_md_free( &ctx );
+
+    return 0;
+}
+
+/**
+ * @brief Verify the manifest via RS256 for the JWS.
+ *
+ * @param pucInput The input over which the RS256 will be verified.
+ * @param ulInputLength The length of \p pucInput.
+ * @param pucSignature The encrypted signature which will be decrypted by \p pucN and \p pucE.
+ * @param ulSignatureLength The length of \p pucSignature.
+ * @param pucN The key's modulus which is used to decrypt \p signature.
+ * @param ulNLength The length of \p pucN.
+ * @param pucE The exponent used for the key.
+ * @param ulELength The length of \p pucE.
+ * @param pucBuffer The buffer used as scratch space to make the calculations. It should be at least
+ * `jwsRSA3072_SIZE` + `jwsSHA256_SIZE` in size.
+ * @param ulBufferLength The length of \p pucBuffer.
+ * @return uint32_t The result of the operation.
+ * @retval 0 if successful.
+ * @retval Non-0 if not successful.
+ */
+static uint32_t prvJWS_RS256Verify( unsigned char * pucInput,
+                                    uint32_t ulInputLength,
+                                    unsigned char * pucSignature,
+                                    uint32_t ulSignatureLength,
+                                    unsigned char * pucN,
+                                    uint32_t ulNLength,
+                                    unsigned char * pucE,
+                                    uint32_t ulELength,
+                                    unsigned char * pucBuffer,
+                                    uint32_t ulBufferLength )
+{
+    AzureIoTResult_t xResult;
+    int32_t lMbedTLSResult;
+
+    if( ulBufferLength < jwsRSA3072_SIZE + jwsSHA256_SIZE )
+    {
+        LogInfo( ( "Buffer Not Large Enough\n" ) );
+        return 1;
```

Comment:
I don't know if we should be returning mbed errors ourselves? Is that something generally done?

Reply:
Having errors come from multiple sources - like sometimes MBEDtls, sometimes 1, sometimes some other sys thing - and here specifically at the `JWS_ManifestAuthenticate` level though the callees matter too - can be confusing.

I think maybe we should just translate - as best we can since there certainly won't be 1-1 mappings - with the `AzureIoTResult_t` code.  This would flow up through system cleaner.  We'd log the mbed error, and probably a lot of times may have to translate into some generic failure. 

We could also I guess extend out `AzureIoTResult_t` with a few cases if interesting enough and we thought caller may need it, though not with 1000's of possible errors.